### ISSUE TITLE
isinstance tuple issue under python 3.6 but not python 3.8 (cv2.VideoCapture)?

### ISSUE THREAD

**Issue by flyingfalling** [Breakthrough/PySceneDetect#220]
I ran an identical script on my laptop (ubuntu 20.04 with python3.8) versus a server (ubuntu 18.04 with python3.6).

I noticed that in new scenedetect 0.5.5, changes were made to VideoManager relating to behaving like cv2.VideoCapture. So this may be related to that.

On the laptop the script runs with no issue. However, on the server (with identical inputs/outputs), it returns the following error (related to python3-opencv?). In the script, I have not accidentally named something cv2.VideoCapture (i.e. nuked the type name with a local variable), which was my first thought. Did type handling change between 3.6 and 3.8?

I will look into this more later, for now, I wanted to see if there was a known issue (google did not turn anything up).

Traceback (most recent call last):
  File "impls/i_02_split_scenes_single.py", line 192, in <module>
    split_scenes_for_video( vid_path, out_path, contthresh, minlen_sec, maxlen_sec );
  File "impls/i_02_split_scenes_single.py", line 137, in split_scenes_for_video
    times, frames = scenes_in_threshold( myvideo_manager, myscene_manager, mybase_timecode, contdet, minlen_sec, maxlen_sec, threshcuts_list );
  File "impls/i_02_split_scenes_single.py", line 54, in scenes_in_threshold
    myscenes1 = detectit( vidman, sceneman, basetime, detector );
  File "impls/i_02_split_scenes_single.py", line 46, in detectit
    vidman.start();
  File "/home/riveale/.local/lib/python3.6/site-packages/scenedetect/video_manager.py", line 606, in start
    self.seek(self._start_time)
  File "/home/riveale/.local/lib/python3.6/site-packages/scenedetect/video_manager.py", line 629, in seek
    if isinstance(self._curr_cap, cv2.VideoCapture):
TypeError: isinstance() arg 2 must be a type or tuple of types

**Comment by flyingfalling** [comment]
This appears to be a bug...I have confirmed a simple fix:

Fixing the line 629 of video_manager.py:
FROM:
if isinstance(self._curr_cap, cv2.VideoCapture):
TO:
if isinstance(self._curr_cap, type(cv2.VideoCapture)):

Fixes the issue.

**Comment by flyingfalling** [comment]
Note: I was running version 0.5.5 installed from PIP.

I checked the most recent git version, it is the same.

**Comment by flyingfalling** [comment]
I'd really prefer not to write a pull request for such a small thing.. ;/ I will when I have time haha.

**Comment by flyingfalling** [comment]
Does anyone know if this is a difference between python 3.6 and 3.8? 3.8 allows you to pass raw class without converting it to "type" format to isinstance, but 3.6 requires the type (or tuple?)?

**Comment by Breakthrough** [comment]
Hmm, interesting - there are unit tests which cover the seek method ( https://github.com/Breakthrough/PySceneDetect/blob/master/tests/test_video_manager.py#L180-L200 ), and the builds on Travis.CI cover both Python 3.6 and 3.8 on Ubuntu 18.  Could you share some more of your script?  Thanks!

**Comment by Breakthrough** [comment]
Also, when you do see the failure, could you add a print statement above it?  e.g. `print(cv2.VideoCapture)` before the `isinstance` line, and provide the output.

**Comment by pkkm** [comment]
I think this problem is likely related to the `opencv-python` version rather than to the Python version. I ran into it when I installed PySceneDetect with `sudo pip3 install scenedetect`, which meant that it used the system version of `opencv-python` (`3.2.0+dfsg-6`) from the `python3-opencv` Debian package. It resolved when I ran `sudo pip3 install 'scenedetect[opencv]'`, which installed a newer version of `opencv-python` from PyPI.

@flyingfalling, what are your `opencv-python` versions on the laptop and server? You can check like this:

```
$ python3
Python 3.7.3 (default, Jan 22 2021, 20:04:44) 
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cv2
>>> cv2.__version__
'4.5.2'
```

**Comment by Breakthrough** [comment]
@pkkm which version of Debian are you running?  I'd like to try and reproduce this on my end.  If you could also show the output of the following with the failing package, it would help as well:

    import cv2
    print(cv2.__version__)
    print(cv2.VideoCapture)

Thank you!

**Comment by pkkm** [comment]
@Breakthrough, I'm running Debian stable (buster). I've just reproduced the problem in a fresh VM, here's the terminal transcript:

```
$ sudo apt-get install python3-opencv
[snip]
Setting up python3-opencv (3.2.0+dfsg-6) ...
[snip]
$ sudo pip3 install scenedetect
Collecting scenedetect
  Downloading https://files.pythonhosted.org/packages/ec/d2/de12be595765235bfbab66d111db7dd390e334dbf4b60986e41003dc9c0b/scenedetect-0.5.5-py2.py3-none-any.whl (107kB)
    100% |████████████████████████████████| 112kB 970kB/s
Collecting tqdm (from scenedetect)
  Downloading https://files.pythonhosted.org/packages/42/d7/f357d98e9b50346bcb6095fe3ad205d8db3174eb5edb03edfe7c4099576d/tqdm-4.61.0-py2.py3-none-any.whl (75kB)
    100% |████████████████████████████████| 81kB 740kB/s
Collecting Click (from scenedetect)
  Downloading https://files.pythonhosted.org/packages/76/0a/b6c5f311e32aeb3b406e03c079ade51e905ea630fc19d1262a46249c1c86/click-8.0.1-py3-none-any.whl (97kB)
    100% |████████████████████████████████| 102kB 1.3MB/s
Requirement already satisfied: numpy in /usr/lib/python3/dist-packages (from scenedetect) (1.16.2)
Collecting importlib-metadata; python_version < "3.8" (from Click->scenedetect)
  Downloading https://files.pythonhosted.org/packages/76/e8/055ba4cbb4662551d7cd2a2193db201e7f597170793b07d75604a06b1074/importlib_metadata-4.3.0-py3-none-any.whl
Collecting typing-extensions>=3.6.4; python_version < "3.8" (from importlib-metadata; python_version < "3.8"->Click->scenedetect)
  Downloading https://files.pythonhosted.org/packages/2e/35/6c4fff5ab443b57116cb1aad46421fb719bed2825664e8fe77d66d99bcbc/typing_extensions-3.10.0.0-py3-none-any.whl
Collecting zipp>=0.5 (from importlib-metadata; python_version < "3.8"->Click->scenedetect)
  Downloading https://files.pythonhosted.org/packages/0f/8c/715c54e9e34c0c4820f616a913a7de3337d0cd79074dd1bed4dd840f16ae/zipp-3.4.1-py3-none-any.whl
Installing collected packages: tqdm, typing-extensions, zipp, importlib-metadata, Click, scenedetect
Successfully installed Click-8.0.1 importlib-metadata-4.3.0 scenedetect-0.5.5 tqdm-4.61.0 typing-extensions-3.10.0.0 zipp-3.4.1
$ wget 'https://github.com/Breakthrough/PySceneDetect/raw/resources/tests/goldeneye/goldeneye.mp4'
[snip]
$ scenedetect --input goldeneye.mp4 detect-content list-scenes
[PySceneDetect] PySceneDetect v0.5.5
[PySceneDetect] Loaded 1 video, framerate: 23.98 FPS, resolution: 1280 x 544
[PySceneDetect] Downscale factor set to 5, effective resolution: 256 x 108
[PySceneDetect] Scene list CSV file name format:
  $VIDEO_NAME-Scenes.csv
Traceback (most recent call last):
  File "/usr/local/bin/scenedetect", line 10, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.7/dist-packages/scenedetect/__main__.py", line 58, in main
    cli.main(obj=cli_ctx)   # Parse CLI arguments with registered callbacks.
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 1072, in main
    ctx.exit()
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 466, in __exit__
    self.close()
  File "/usr/local/lib/python3.7/dist-packages/click/core.py", line 595, in close
    self._exit_stack.close()
  File "/usr/lib/python3.7/contextlib.py", line 519, in close
    self.__exit__(None, None, None)
  File "/usr/lib/python3.7/contextlib.py", line 511, in __exit__
    raise exc_details[1]
  File "/usr/lib/python3.7/contextlib.py", line 496, in __exit__
    if cb(*exc_details):
  File "/usr/lib/python3.7/contextlib.py", line 383, in _exit_wrapper
    callback(*args, **kwds)
  File "/usr/local/lib/python3.7/dist-packages/scenedetect/cli/context.py", line 239, in process_input
    self.video_manager.start()
  File "/usr/local/lib/python3.7/dist-packages/scenedetect/video_manager.py", line 606, in start
    self.seek(self._start_time)
  File "/usr/local/lib/python3.7/dist-packages/scenedetect/video_manager.py", line 629, in seek
    if isinstance(self._curr_cap, cv2.VideoCapture):
TypeError: isinstance() arg 2 must be a type or tuple of types
$ python3
Python 3.7.3 (default, Jan 22 2021, 20:04:44)
[GCC 8.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import cv2
>>> cv2.__version__
'3.2.0'
>>> cv2.VideoCapture
<built-in function VideoCapture>
```

**Comment by Breakthrough** [comment]
Thanks for providing the output and confirming the issue!  I was worried about having to write a wrapper in `platform.py`, but after another look over the VideoManager, that if statement is actually not necessary.  I'll fix this up for the next release and add it to the list of known issues.

Thank you!

**Comment by Breakthrough** [comment]
Fixed in v0.5.6 branch, will be included in next release of PySceneDetect.  Thank you for the submission!
